Add inbuilt Opengl renderer to window

emmauss opened the pull request:
We can now ditch the extra window. Allows rendering in the main window. the old GLScreen is still there, but can not be accessed anymore through. Will remove it if its decided its no longer necessary.
So far, I've not noticed any slowdowns. the rendering logic is more or less the same, and its multithreaded. Contrary to what I thought before, you can modify the underlying context in gtk, with it crashing on you.

This PR is ready for review.

![Screenshot_2020-02-06 Screenshot from 2020-02-06 11-38-33 png (PNG Image, 1920 × 1080 pixels) - Scaled (83%)](https://user-images.githubusercontent.com/5307160/73933999-8aac4100-48d5-11ea-9f81-829afd1ce1d8.png)
![Screenshot_2020-02-06 Screenshot from 2020-02-06 11-38-52 png (PNG Image, 1920 × 1080 pixels) - Scaled (83%)](https://user-images.githubusercontent.com/5307160/73934005-8d0e9b00-48d5-11ea-96f4-a855f36f35a3.png)

emmauss commented:
as an aside, code for the widget is here, https://gist.github.com/emmauss/15fef952fe25d7f838ccd6a5f438249b

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+
+        public bool IsActive   { get; set; }
+        public bool IsStopped  { get; set; }
+        public bool IsFocussed { get; set; } = false;
```
```suggestion
        public bool IsFocused { get; set; }
```
A bool is always false by default.

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+
+        private Input.NpadController _primaryController;
+
+        public GLRenderer(Switch device) :base (new GraphicsMode(), 3, 3, GraphicsContextFlags.ForwardCompatible)
```
```suggestion
        public GLRenderer(Switch device) : base (new GraphicsMode(), 3, 3, GraphicsContextFlags.ForwardCompatible)
```
But maybe you can add the base part on a new line?

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+
+        protected override bool OnConfigureEvent(EventConfigure evnt)
+        {
+            var result  = base.OnConfigureEvent(evnt);
```
```suggestion
            var result = base.OnConfigureEvent(evnt);
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+            // OpenTK always captures mouse events, even if out of focus, so check if window is focused.
+            if (IsFocussed && _mousePressed)
+            {
+                int scrnWidth = AllocatedWidth;
```
```suggestion
                int screenWidth = AllocatedWidth;
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+            if (IsFocussed && _mousePressed)
+            {
+                int scrnWidth = AllocatedWidth;
+                int scrnHeight = AllocatedHeight;
```
```suggestion
                int screenHeight = AllocatedHeight;
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+                    _mouseX < endX &&
+                    _mouseY < endY)
+                {
+                    int scrnMouseX = (int)_mouseX - startX;
```
```suggestion
                    int screenMouseX = (int)_mouseX - startX;
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+                    _mouseY < endY)
+                {
+                    int scrnMouseX = (int)_mouseX - startX;
+                    int scrnMouseY = (int)_mouseY - startY;
```
```suggestion
                    int screenMouseY = (int)_mouseY - startY;
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+                    int scrnMouseX = (int)_mouseX - startX;
+                    int scrnMouseY = (int)_mouseY - startY;
+
+                    int mX = (scrnMouseX * TouchScreenWidth) / scrnWidth;
```
```suggestion
                    int mouseX = (screenMouseX * TouchScreenWidth) / scrnWidth;
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,422 @@
+                    int scrnMouseY = (int)_mouseY - startY;
+
+                    int mX = (scrnMouseX * TouchScreenWidth) / scrnWidth;
+                    int mY = (scrnMouseY * TouchScreenHeight) / scrnHeight;
```
```suggestion
                    int mouseY = (screenMouseY * TouchScreenHeight) / scrnHeight;
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -43,32 +43,34 @@ public class MainWindow : Window
-        [GUI] LevelBar      _progressBar;
-        [GUI] MenuItem      _openDebugger;
-        [GUI] MenuItem      _toolsMenu;
+
```
Extra empty line

emmauss commented:
addressed comments

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+
+        public bool IsActive   { get; set; }
+        public bool IsStopped  { get; set; }
+        public bool IsFocused { get; set; } = false;
```
```suggestion
        public bool IsFocused  { get; set; }
```
A bool is already false by default and brackets aren’t aligned.

gdkchan commented:
> with it crashing on you.
I guess this is a typo?
> the old GLScreen is still there, but can not be accessed anymore through. Will remove it if its decided its no longer necessary.
I think it should be removed.

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+
+        private Input.NpadController _primaryController;
+
+        public GLRenderer(Switch device) :
```
It could be nice to made the same alignment as here: https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx/Ui/GLScreen.cs#L42

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+
+            _primaryController = new Input.NpadController(ConfigurationState.Instance.Hid.JoystickControls);
+
+            AddEvents((int)(Gdk.EventMask.ButtonPressMask |
```
Gdk alignment is weird here

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+
+        public void HandleScreenState(KeyboardState keyboard)
+        {
+            bool toggleFullscreen = keyboard.IsKeyDown(OpenTK.Input.Key.F11) ||
```
The || need to be on the next line if you put the && at the begining

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+
+            if (IsFocused)
+            {
+                if (this.ParentWindow.State.HasFlag(Gdk.WindowState.Fullscreen) )
```
```suggestion
                if (this.ParentWindow.State.HasFlag(Gdk.WindowState.Fullscreen))
```

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+            WaitEvent.Set();
+        }
+
+        protected override bool OnConfigureEvent(EventConfigure evnt)
```
```suggestion
        protected override bool OnConfigureEvent(EventConfigure event)
```
And same in others functions

emmauss replied:
event is a reserved word

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+                        this.ParentWindow.Title = _newTitle;
+                    });
+                }
+                if (IsFocused)
```
Add a new line before

AcK77 commented on Ryujinx/Ui/GLRenderer.cs:
```diff
@@ -0,0 +1,489 @@
+                {
+                    screenHeight = (AllocatedWidth * TouchScreenHeight) / TouchScreenWidth;
+                }
+
```
Extra empty line

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -30,7 +29,8 @@ public class MainWindow : Window
         private static HLE.Switch _emulationContext;
 
-        private static GlScreen _screen;
+        // private static GlScreen _screen;
```
If you don’t use it, you can remove it, or at least leave a comment on why this line is commented!

Xpl0itR commented on Ryujinx/Ui/ScopedGLContext.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Ryujinx.Ui
+{
+    class ScopedGLContext : IDisposable
```
```suggestion
    class ScopedGlContext : IDisposable
```

Xpl0itR commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -30,7 +29,8 @@ public class MainWindow : Window
 
-        private static GlScreen _screen;
+        // private static GlScreen _screen;
+        private static GLRenderer _gLWigdet;
```
```suggestion
        private static GLRenderer _gLWidget;
```

Xpl0itR commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -43,32 +43,38 @@ public class MainWindow : Window
+        [GUI] CheckMenuItem  _favToggle;
+        [GUI] MenuItem       _firmwareInstallFile;
+        [GUI] MenuItem       _firmwareInstallDirectory;
         [GUI] MenuItem      _openDebugger;
```
```suggestion
        [GUI] MenuItem       _openDebugger;
```

Xpl0itR commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -384,26 +390,81 @@ private void CreateGameWindow(HLE.Switch device)
                 _stopEmulation.Sensitive            = false;
                 _firmwareInstallFile.Sensitive      = true;
                 _firmwareInstallDirectory.Sensitive = true;
+
```
remove empty line

Xpl0itR commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -384,26 +390,81 @@ private void CreateGameWindow(HLE.Switch device)
+            Application.Invoke(delegate
+            {
+                _viewBox.Remove(_gLWigdet);
+                _gLWigdet.Dispose();
```
Crashes here when disposing the widget when stopping emulation, only happens sometimes.

emmauss commented:
>     * nit: If you open MainWindow.glade in glade you will see that the footer box has spaces for 4 child widgets but only has 3 child widgets.
Done
>     * The custom gl widget has some random console logging. https://gist.github.com/emmauss/15fef952fe25d7f838ccd6a5f438249b#file-glwidget-cs-L278
Will remove it in an update to the wigdet package
>     * Resizing and fullscreening the window is bugged if the window is not being rendered to
Noting I can do about this. I've tried clearing the screen after resizing. it did not work.

Addressed comments

gdkchan commented on Ryujinx/Ui/ScopedGlContext.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        private IGraphicsContext _graphicsContext;
+
+        private static readonly object _lock = new object();
```
Why is this lock static?
